libvast/builtins/operators/top_rare.cpp

```diff
@@ -0,0 +1,86 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <vast/argument_parser.hpp>
+#include <vast/concept/parseable/string/char_class.hpp>
+#include <vast/concept/parseable/vast/pipeline.hpp>
+#include <vast/detail/string_literal.hpp>
+#include <vast/error.hpp>
+#include <vast/location.hpp>
+#include <vast/logger.hpp>
+#include <vast/pipeline.hpp>
+#include <vast/plugin.hpp>
+
+namespace vast::plugins::top_rare {
+
+namespace {
+
+template <detail::string_literal Name, detail::string_literal SortOrder>
+class top_rare_plugin final : public virtual operator_parser_plugin {
+  auto name() const -> std::string override {
+    return std::string{Name.str()};
+  }
+
+  auto parse_operator(parser_interface& p) const -> operator_ptr override {
+    auto parser = argument_parser{
+      std::string{Name.str()}, fmt::format("https://docs.tenzir.com/docs/next/"
+                                           "operators/transformations/{}",
+                                           Name.str())};
+    auto field = located<std::string>{};
+    auto count_field = std::optional<located<std::string>>{};
+    parser.add(field, "<str>");
+    parser.add("-c,--count-field", count_field, "<str>");
+    parser.parse(p);
+    if (count_field) {
+      if (count_field->inner.empty()) {
+        diagnostic::error("need a string value for `count-field` parameter")
+          .primary(field.source)
+          .throw_();
+      }
+      if (count_field->inner == field.inner) {
+        diagnostic::error("invalid duplicate field value `{}` for count and "
+                          "value fields",
+                          field.inner)
+          .primary(field.source)
+          .secondary(count_field->source)
+          .throw_();
+      }
+    } else {
+      if (field.inner == default_count_field) {
+        diagnostic::error("invalid duplicate field value `{}` for count and "
+                          "value fields",
+                          field.inner)
+          .primary(field.source)
+          .throw_();
+      } else {
+        count_field.emplace();
+        count_field->inner = default_count_field;
+      }
+    }
+    // TODO: Replace this textual parsing with a subpipeline to improve
+    // diagnostics for this operator.
+    auto repr = fmt::format("summarize "
+                            "{0}=count({1}) by "
+                            "{1} | sort {0} {2}",
+                            count_field->inner, field.inner, SortOrder.str());
+    return std::move(*pipeline::internal_parse_as_operator(repr));
```

Comment:
The `internal_parse` call can fail here (for example with `top "x y z"`, I think).

Reply:
I thought `internal_parse_as_operator` would throw an error, too.

Are `internal` methods exempt from using `diagnostics`?

Reply:
No, it returns a `caf::expected`, which contains a stringified version of the diagnostic (this is to make the transition to diagnostics easier). This API will go away eventually, it only is a temporary replacement for the `pipeline::parse(...)`.

Reply:
Perhaps we also want to have a version of this which emits diagnostics instead, but displaying these diagnostics will only contain limited information, as all `location` values will be `location::unknown` (because it does not come from user input).